### PR TITLE
internal: sync from downstream

### PR DESCRIPTION
**Pull request by lnicola**

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -48,6 +48,7 @@ jobs:
       CC: deny_c
       RUST_CHANNEL: "${{ needs.changes.outputs.proc_macros == 'true' && 'nightly' || 'stable' }}"
       USE_SYSROOT_ABI: "${{ needs.changes.outputs.proc_macros == 'true' && '--features sysroot-abi' || '' }}"
+      RUSTUP_RUSTC_DEV: "${{ needs.changes.outputs.proc_macros == 'true' && 'rustc-dev' || '' }}"
```
Comment (lnicola) [inline .github/workflows/ci.yaml:51]: We want this, right?

Comment (Veykril) [reply .github/workflows/ci.yaml:51]: No we shouldn't need this 🤔 our CI should not rely on the rustc crates

File: crates/proc-macro-srv/src/lib.rs [hunk 2]
```diff
@@ -11,11 +11,12 @@
 //!   rustc rather than `unstable`. (Although in general ABI compatibility is still an issue)…
 
 #![cfg(any(feature = "sysroot-abi", rust_analyzer))]
-#![feature(proc_macro_internals, proc_macro_diagnostic, proc_macro_span)]
+#![feature(proc_macro_internals, proc_macro_diagnostic, proc_macro_span, rustc_private)]
 #![warn(rust_2018_idioms, unused_lifetimes)]
 #![allow(unreachable_pub, internal_features)]
 
 extern crate proc_macro;
+extern crate rustc_driver as _;
```
Comment (Veykril) [inline crates/proc-macro-srv/src/lib.rs:19]: This is missing a cfg `#[cfg(feature = "in-rust-tree")]`

File: crates/proc-macro-srv/src/lib.rs [hunk 3]
```diff
@@ -11,11 +11,12 @@
 //!   rustc rather than `unstable`. (Although in general ABI compatibility is still an issue)…
 
 #![cfg(any(feature = "sysroot-abi", rust_analyzer))]
-#![feature(proc_macro_internals, proc_macro_diagnostic, proc_macro_span)]
+#![feature(proc_macro_internals, proc_macro_diagnostic, proc_macro_span, rustc_private)]
```
Comment (Veykril) [inline crates/proc-macro-srv/src/lib.rs:14]: This feature needs to be gated `#[cfg_attr(feature = "in-rust-tree", feature(rustc_private)]`

File: crates/proc-macro-srv/src/lib.rs [hunk 4]
```diff
@@ -11,11 +11,14 @@
 //!   rustc rather than `unstable`. (Although in general ABI compatibility is still an issue)…
 
 #![cfg(any(feature = "sysroot-abi", rust_analyzer))]
+#![cfg_attr(feature = "in-rust-tree", feature(rustc_private))]
```
Comment (lnicola) [inline crates/proc-macro-srv/src/lib.rs:14]: Like this, right?

**Comment by Veykril** [comment]
Thanks for doing the syncs as always, especially since they seem to have been more trouble the past weeks 😅

File: crates/proc-macro-srv/src/lib.rs (hunk 4 above)
Comment (Veykril) [reply crates/proc-macro-srv/src/lib.rs:14]: Yes

**Comment by Veykril** [comment]
CI failure is due to a new lint on nightly fwiw, since this touches the proc-macro server source CI runs the nightly toolchain. You can probably just pick the changes from this file here https://github.com/rust-lang/rust-analyzer/pull/16309/files#diff-0449230bb140c3330091dc8be2f8b5d227e6875704429de5116f428fcbe7b532

**Comment by Veykril** [comment]
🤨 I can't repro that, and nothing should be causing that here, might be spurious let's see
@bors r+

**Comment by Veykril** [comment]
Hm, feel free to just cfg that test out for now (since ignoring is rejected by CI as well), I'll look into what's going on there later. Probably related to the dylib linking changes in the sysroot, fairly sure the error there is incorrect and its something else.

**Comment by lnicola** [comment]
I can reproduce the failure using `RUN_SLOW_TESTS=1 cargo test -p rust-analyzer --test slow-tests --features sysroot-abi resolve_proc_macro`.

File: crates/rust-analyzer/tests/slow-tests/main.rs [hunk 5]
```diff
@@ -835,7 +835,7 @@ fn main() {
 #[cfg(any(feature = "sysroot-abi", rust_analyzer))]
 fn resolve_proc_macro() {
     use expect_test::expect;
-    if skip_slow_tests() {
+    if skip_slow_tests() || true {
```
Comment (lnicola) [inline crates/rust-analyzer/tests/slow-tests/main.rs:838]: Added an early exit here.

**Comment by Veykril** [comment]
Ugh right, I forgot to set the slow test var, repros for me as well then, that's good. Will take a look at that in a bit.

**Comment by Veykril** [comment]
Fairly certain the cause is https://github.com/rust-lang/rust-analyzer/pull/16307 (which didn't run the test)

@bors r+

**Comment by lnicola** [comment]
> Thanks for doing the syncs as always, especially since they seem to have been more trouble the past weeks

Months, you mean :smile:.

> Fairly certain the cause is https://github.com/rust-lang/rust-analyzer/pull/16307 (which didn't run the test)

Yeah, reverting that fixes the failure :-(.

**Comment by Veykril** [comment]
Ye okay I know why, now that we no longer save the document in the vfs we can't filter out `set_file_content` calls that don't change the document value, and something seems to report changing the file content multiple times hence the failure.

**Comment by lnicola** [comment]
```
2024-01-09T08:54:56.688606Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/bar/src/lib.rs)
2024-01-09T08:54:56.688821Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/foo/src/main.rs)
2024-01-09T08:54:56.689274Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/bar/Cargo.toml)
2024-01-09T08:54:56.689285Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/bar/src/lib.rs)
2024-01-09T08:54:56.689294Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/foo/Cargo.toml)
2024-01-09T08:54:56.689300Z  INFO vfs: set_file_contents(/tmp/testdir/87771_0/foo/src/main.rs)
```

CC https://github.com/rust-lang/rust-analyzer/issues/14730?

**Comment by Veykril** [comment]
Yes, that pretty much (and another smaller issue).

**Pull request merged**